### ISSUE TITLE
emptyDir persistent volumes

### ISSUE THREAD

**Issue by adampl** [kubernetes/kubernetes#75378]
**What would you like to be added**:
I need a volume accessible from different pods on each node, serving as a **shared file cache**. In my case it should be a `tmpfs` volume to provide maximum speed, so I'd use `emptyDir` with `medium: Memory`. 

Therefore I'd like `emptyDir` to be available in `PersistentVolumeSpec` just like all other volume sources.

**Why is this needed**:
Currently (AFAIK) the only way to achieve this is to:
- `mkdir` on each node's host filesystem at some predefined path
- optionally mount a `tmpfs` volume there
- create a `PersistentVolume` of type `hostPath` or `local` to mount that path

This method seems very "un-kubernetic" to me and could be easily automated.
Even pure Docker can do this:
```
docker volume create --driver local --opt type=tmpfs --opt device=tmpfs \
       --opt o=size=100m,uid=1000 foo
```

There is an old similar issue #27732 but I don't think it has to be limited to `tmpfs` only.

**Comment by adampl** [comment]
@kubernetes/sig-storage-feature-requests

**Comment by OJFord** [comment]
My use-case is not tmpfs; this is an issue whenever a pod provides a fuse filesystem backed by physical disk on the node, for other pods to consume.

Currently, in order to use it in another pod it needs to be `local`/`hostPath` (or some other PV), but there's really no need to have to create a mountpoint on the node for it other than this; as OP says, doing so
> seems very "un-kubernetic"

**Comment by msau42** [comment]
I think this may be possible with CSI ephemeral volumes now. I believe @kfox1111 had a similar use case.

**Comment by kfox1111** [comment]
There is no pv support, but you can use the csi ephemeral inline support to write drivers that can do this I think.

**Comment by adampl** [comment]
@kfox1111 But, if I use the inline volume, how can I share the same volume between different pods?

**Comment by kfox1111** [comment]
The ephemeral inline spec only states that the driver will get a unique volumeHandle following the pod lifecycle. It doesn't restrict you to what you can do with the handle. Used along with the pod info on mount feature (https://kubernetes-csi.github.io/docs/pod-info.html), you could probably make a directory on the filesystem based on the csi.storage.k8s.io/pod.namespace attribute instead of volumeHande, reference count using volumeHandle, and remove the namespace directory when refcount goes to zero.

You could even add a namespace + name thing so you could have multiple per namespace. The api could look like:
```
  volumes:
  - name: data
    csi:
      driver: namespaced.emptydir.csi.k8s.io
      volumeAttributes:
          name: cache
```
Could map to /var/lib/namespacedemptydir/$namespace/$name

**Comment by kfox1111** [comment]
@msau42 That driver I think would probably be another good candidate for living under github.com/kubernetes-csi/csi-driver-*, as its in no way vendor specific.

**Comment by msau42** [comment]
It can potentially go there if there are folks willing to maintain and manage the releases and testing

**Comment by kfox1111** [comment]
I can help with that. but I don't have time at the moment to write the code. But can help give pointers if @adampl is interested.

**Comment by adampl** [comment]
Well, I think CSI is for external storage services. Inline `emptyDir` volumes and `hostPath/local` PVs are built-in features, so involving the whole CSI framework looks like a serious overkill - compared to slightly modifying the built-in `emptyDir` to make it PV compatible. Unfortunately, I'm not a Go/k8s developer, so I'd have to dig into it for a long time, not to mention writing a KEP or getting it approved.

@msau42 Isn't this sample Hostpath driver the kind of thing I'm looking for? https://github.com/kubernetes-csi/csi-driver-host-path but it's not production ready :(

**Comment by kfox1111** [comment]
ephemeral csi was written specifically to address the growing need for extra types of enhanced emptyDir usecases such as emptyDir persistent volumes, secret injection, data only containers, etc. If all drivers were added directly to kubelet, it would get unwieldy. emptydir and hostpath in kubelet are enough to get csi going and the rest can go out of tree. I do think its the right tool for the job.

yeah, that hostpath driver would need a lot of work I think to be usable in that form I think.

Lets keep this issue open. I think there is enough detail here for someone to implement it if they want to take a stab at it, and it is a useful thing to get implemented.

**Comment by msau42** [comment]
Yeah I don't think the hostpath driver should be enhanced. It's mainly used for our testing purposes and is more of a mock driver than a real driver.  I was thinking a brand new driver but yes, it will require some investment from the community to maintain and test.

**Comment by adampl** [comment]
@kfox1111 But `emptyDir` is already built-in... and I bet the PV feature is pretty simple to implement for someone who knows the API and lightweight to manage.

@msau42 Do you mean the built-in HostPath driver, or the CSI plugin?

Anyway, I don't get why the built-in HostPath is considered inferior (https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes). It's up to me as an admin/architect to take into account the specific nature of this volume. Forcing people to use pod-local mounts makes things worse, not better.

**Comment by kfox1111** [comment]
hostPath and emptydir are built in. If every possible emptydir style driver was in tree it would be an unwieldy mess of a code base. There are tons of drivers of that kind. coming up with a solid plugin mechanism resolves the issue.

The other issue is you must wait for several releases to get new code into kubelet, this can take years before an operator can actually start using it. Starting with 1.16, drivers can be written in the future and still work with that version. If you want something that could work with a currently released kubernetes, this is your only path.

The csi hostpath driver really should be renamed IMO. Its only intended for testing, and behaves significantly different then the in tree hostpath driver. Being named so similarly has caused many miscommunications. @msau42 is that something we should do?

Writing a driver to do the above feature I think would be pretty simple. It would be, fork this repo: 
https://github.com/kubernetes-csi/csi-driver-image-populator/

replace all image-populator strings with a different name

And edit the two functions:
NodePublishVolume
NodeUnpublishVolume

to do:
open file at 
/var/lib/<drivername>/<podnamespace>/<volumeAttribute.name>/refcount. set to 0 if non existent.

on publish increment refcount by one. on unpublish decrement by one. If decremented to 0, rm -rf /var/lib/<drivername>/<podnamespace>/<volumeAttribute.name>

On publish, bind mount /var/lib/<drivername>/<podnamespace>/<volumeAttribute.name>/volume
on unpublish, unmount.

I think that would produce a usable driver.

**Comment by msau42** [comment]
Renaming things is difficult and can potentially break a lot of things. I am open to improving documentation though to clarify the hostpath csi driver is only intended for testing purposes.

**Comment by adampl** [comment]
@kfox1111 I think you have mistaken something. There is only one `emptyDir` volume type, and it has only one option - `medium` with only 2 possible values, the default "" (disk) and "Memory" (tmpfs). Of all the multiple in-tree drivers, only the following cannot be used with PVs:

- configMap
- downwardAPI
- emptyDir
- gitRepo (deprecated)
- local
- persistentVolumeClaim
- projected
- secret

Of these, only `emptyDir` and `local` are writable and can be used for local data storage.
I see no reason for excluding these 2 from the "PV club".

**Comment by kfox1111** [comment]
But there is not a driver in existence that does what you want. an emptyDir volume will be deleted as soon as the pod referencing it goes. There is no Persistent Volume in that. So just hooking it up to a PV doesn't work. Once you start  down that path, then there are lots of behavioural options that someone may want to use for emptyDir in a pv. Does it follow the pod lifecycle? namespace lifecycle? node lifecycle, pv lifecycle? Some combination? Are there more then one per namespace per node? Is it shared amongst all namespaces? If you want one combination of options, someone else may need another combination. Each is probably valid. So do we support them all? That would make things unwieldy. If plugable, then specific implementations with all the requirements can be made as the need arises. Maybe if a few of them become very common, then those get merged in?

The local driver does work with pv's. Just not in the way that would facilitate a cache.

**Comment by kfox1111** [comment]
yes, it is unfortunately. Your trying to tie together several constructs that weren't intended to be used together, so there are no established communication/code paths that can easily handle that.

I think you can get close by writing an ephemeral csi driver, minus the pv part.

If you really want the pv part, you would need a full blown csi driver as only they have all the hooks needed from kubernetes to the driver, in particular Controller.DeleteVolume. But its called only to one instance. So some mechanism would be needed to get this DeleteVolume call to all nodes that ever had that backing volume used so they could clean up the volume. There is no current mechanism in place for that. It would be a fair amount of work. The edge cases are difficult too. When do you know the delete happened to all affected nodes so you can garbage collect the notification? The csi driver could do it itself out of band, but then the driver would be specific to Kubernetes (maybe ok) too.

**Comment by davidkarlsen** [comment]
Here is another usecase: https://groups.google.com/forum/#!msg/memcached/faWH90ivfE4/QPVMGSKiBAAJ

**Comment by kfox1111** [comment]
@davidkarlsen I believe an ephemeral csi driver could handle that use case. No enhancement to k8s itself is needed.

**Comment by kfox1111** [comment]
@msau42 I've contemplated writing such a driver before. looks like more folks want such a thing. Could we get another space in github.com/kubernetes-csi/csi-driver-<something> to get contributions going?

**Comment by msau42** [comment]
Do we have people willing to step up and contribute and maintain this driver?  A big problem we have right now in kubernetes-csi is lack of contributors and maintainers for many of the drivers.

**Comment by kfox1111** [comment]
I'd be willing to work on it. Anyone else?

**Comment by hasheddan** [comment]
@kfox1111 @msau42 did anyone ever get started on this?

**Comment by kfox1111** [comment]
No, sorry. It got bumped by other priorities. I still think it could be fairly easily written as a csi ephemeral driver. The image-populator one would be a good starting point I think for someone wanting to try and implement this. The developer would need to:
1. remove the buildah related code
2. add reference counting to cleanup the volume once all relevant pods are removed from the node.
